IAM: Handle s2s token request

reinkrul opened the pull request:
Fixes https://github.com/nuts-foundation/nuts-node/issues/2495

Noteworthy things:
- I chose to return an unencrypted JWT as access token for now; since we don't have persisted token storage and I can't remember what we decided (using a JWT or opaque value that maps to a server-side stored token) -> we now an opaque token
- I assumed we don't need to support `path_nested` in Presentation Exchanges (not sure, maybe only if we support VCs and VPs in JWT format?) -> support pending (https://github.com/nuts-foundation/nuts-node/pull/2563)
- Matching the presentation submission against the Verifiable Presentation (in JSON-LD format) is kinda fiddly, since our VP JSON-LD marshalling compacts arrays. This means path indexers won't work if there's just 1 credential in the VP. I fixed with some `[]interface{}` checking in `PresentationSubmission.Credentials()`, but theoretically there could be presentation definitions that map to compacted arrays in other Verifiable Credential types. So we might need to convert JSON-LD VPs/VCs to plain JSON first (uncompacting arrays). But maybe leave that out of scope for now, and fix it when implementing JWT support? 

Requires https://github.com/nuts-foundation/nuts-node/pull/2566

woutslakhorst commented:
We need to do opaque tokens since we need access to the entire VC/VP, they are too big to put in a header. Introspection will need to get them from session storage.

reinkrul commented:
Rebased, changes:
- Removed access token JWT implementation, replaced secure random code stored in session database
- Returned `OAuth2Error` instead of "normal" Go error

woutslakhorst commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -19,49 +19,157 @@
-	return scope, nil
+	// Check signatures and VC issuer trust
+	issueTime := time.Now()
+	_, err = s.vcr.Verifier().VerifyVP(vp, true, false, &issueTime)
```
`VerifyVP` does not check if the given credentials are owned by the client. RFC021 states:

| 7. The `verificationMethod` field of the proof MUST match the `credentialSubject.id` field from all the Verifiable Credentials that are used to request the access token.

This is not a general requirement for VP validation. We should hardcode it here. Also as discussed in front of the whiteboard, this check must be done before the discovery and cryptographic check to prevent DOS attacks.

woutslakhorst commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -19,49 +19,157 @@
+		Expiration:   issueTime.Add(accessTokenValidity),
+		Presentation: presentation,
+	}
+	err := s.storageInstance.GetSessionDatabase().GetStore(accessTokenValidity, issuer.String(), "accesstoken").Put(accessToken.Token, accessToken)
```
maybe add `s2s` as first key so we can debug how many s2s keys are active?

reinkrul replied:
Yeah, I prefixed the OpenID keys with `openid`, so `s2s` is in order here I think

reinkrul commented:
Further VC/VP verification optimizations are part of https://github.com/nuts-foundation/nuts-node/issues/2528 (since it's existing code).

woutslakhorst commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -100,14 +106,19 @@ func (s Wrapper) handleS2SAccessTokenRequest(issuer did.DID, params map[string]s
-	issueTime := time.Now()
-	_, err = s.vcr.Verifier().VerifyVP(vp, true, false, &issueTime)
+	// Check VP signer == VC subject (presenter is holder of VCs)
+	if vp.Holder == nil {
```
Holder is optional, if not present use kid/verificationMethod

woutslakhorst commented on vcr/credential/util.go:
```diff
@@ -0,0 +1,65 @@
+// If the check succeeds, it returns nil.
+func VerifyPresenterIsHolder(vp vc.VerifiablePresentation) error {
+	// Check VP signer == VC subject (presenter is holder of VCs)
+	if vp.Holder == nil {
```
As mentioned before, use kid/verificationMethod. Holder is used in some weird use-cases only.

gerardsn commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -19,49 +19,180 @@
+	// 2. Take VCs mapped by the presentation submission
+	// 3. Match the VCs from (2) to the definition from (1). This should yield the same list of credentials.
+	//    This actually creates a new submission.
+	presentationDefinition := s.auth.PresentationDefinitions().ByScope(scope)
+	if presentationDefinition == nil {
```
`ByScope` should probably return an error

gerardsn commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -19,49 +19,180 @@
+		}
+	}
+
+	submissionCredentialsJSON, _ := json.Marshal(submissionCredentials)
+	matchedCredentialsJSON, _ := json.Marshal(matchedCredentials)
+	if !bytes.Equal(submissionCredentialsJSON, matchedCredentialsJSON) {
+		return nil, OAuth2Error{
+			Code:        InvalidRequest,
+			Description: "presentation_submission parameter is invalid: re-evaluation of the credentials against the presentation definition yields different credentials (bug or attempted hack)",
+		}
+	}
```
I think this test can fail for several reasons even if `submissionCredentialsJSON` and `matchedCredentialsJSON` contain the same set (different order of VCs, different order of the elements in the JSON). Can we assume that the arrays are the same since `presentationDefinition.Match` did not return an error?

reinkrul replied:
Discussed: we need to introduce a reusable submission validation function

gerardsn commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -19,49 +19,180 @@
+	}
+
+	// Check signatures and VC issuer trust
+	issueTime := time.Now()
+	_, err = s.vcr.Verifier().VerifyVP(vp, true, false, &issueTime)
```
`issuedTime` doesn't make sense here but does later on. Rename to `now`?

gerardsn commented on vcr/credential/util.go:
```diff
@@ -0,0 +1,65 @@
+		return err
+	}
+	if *vp.Holder != credentialSubjectID.URI() {
+		return errors.New("not all VC credentialSubject.id match VP holder")
```
ResolveSubjectDID confirmed that all VCs have the same subjectDID, so its all or nothing here.

gerardsn reviewed (changes requested):
The presentation submission is incomplete without `path_nested`, but I'm not sure if that is blocking now.

[OpenID4VP §6.1](https://openid.net/specs/openid-4-verifiable-presentations-1_0.html#name-response-parameters) has the following example of a presentation submission for a VP with nested credentials that matches our use case

```
{
    "id": "Presentation example 1",
    "definition_id": "Example with selective disclosure",
    "descriptor_map": [
        {
            "id": "ID card with constraints",
            "format": "ldp_vp",
            "path": "$",
            "path_nested": {
                "format": "ldp_vc",
                "path": "$.verifiableCredential[0]"
            }
        }
    ]
}
```

but I think JSONPath/PEX also allows this to be expressed as

```
{
    "id": "Presentation example 1",
    "definition_id": "Example with selective disclosure",
    "descriptor_map": [
        {
            "id": "ID card with constraints",
            "format": "ldp_vc",
            "path": "$.verifiableCredential[0]"
        }
    ]
}
```

IMO the former is nicer since you get the format of the VP, but not strictly necessary. Maybe we can get away with skipping `path_nested` for now?

reinkrul commented on vcr/pe/submission.go:
```diff
@@ -0,0 +1,73 @@
+	var credentials []vc.VerifiableCredential
+	var mappedCredentials = make(map[string]bool, len(s.DescriptorMap))
+	for _, inputDescriptor := range s.DescriptorMap {
+		// Note: path_nested is not supported. Do we need it?
```
For now, `path` suffices. Will create an issue to support path_nested

reinkrul commented on vcr/pe/test.go:
```diff
@@ -0,0 +1,73 @@
+	  }
+	}
+  ],
+  "format": {
```
Not used atm, will break when it becomes relevant

reinkrul commented:
@woutslakhorst @gerardsn rewritten most of PR based on introduced `PresentationSubmission.Validate()` function and resolved earlier feedback.

gerardsn commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -40,40 +41,106 @@ import (
+// 2. Check VC issuer trust
+// 3. Check signer of VP == subject of VCs
+// 4. Check signatures of VP and VCs
+func (s Wrapper) handleS2SAccessTokenRequest(issuer did.DID, params map[string]string) (HandleTokenRequestResponseObject, error) {
```
missing checks for
- replay prevention (per VP, since we don't provide a nonce that must be used)
- validity period: (exp-nbf <= 5s)
- audience

gerardsn commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -40,40 +41,106 @@ import (
-	return nil, nil
+	// Check signatures of VP and VCs. Trust should be established by the Presentation Definition.
+	for _, presentation := range pexEnvelope.Presentations {
+		_, err = s.vcr.Verifier().VerifyVP(presentation, true, true, nil)
```
`VerifyVP(..., validAt)` should not be nil?

reinkrul replied:
Defaults to current time

reinkrul commented on vcr/pe/util.go:
```diff
@@ -0,0 +1,87 @@
+	"github.com/nuts-foundation/go-did/vc"
+	"strings"
+)
+
+type Envelope struct {
+	Interface     interface{}
+	Presentations []vc.VerifiablePresentation
+}
```
I'll make a separate PR

reinkrul replied:
See https://github.com/nuts-foundation/nuts-node/pull/2620

reinkrul commented on vcr/pe/util.go:
```diff
@@ -0,0 +1,87 @@
+// ParseEnvelope parses a Presentation Exchange envelope, which is a JSON type that encompasses zero or more Verifiable Presentations.
+// It returns the envelope as interface{} (for use in PresentationSubmission.Validate() and PresentationSubmission.Resolve()),
+// and as parsed Verifiable Presentations.
+func ParseEnvelope(envelopeBytes []byte) (*Envelope, error) {
```
Fixed in https://github.com/nuts-foundation/nuts-node/pull/2620

reinkrul commented on auth/services/oauth/relying_party.go:
```diff
@@ -166,15 +166,15 @@ func (s *relyingParty) RequestRFC021AccessToken(ctx context.Context, requester d
 		return nil, core.Error(http.StatusPreconditionFailed, "no matching credentials")
 	}
-	expires := time.Now().Add(time.Minute * 15) //todo
-	nonce := nutsCrypto.GenerateNonce()
```
It was used for `challenge`, which is incorrect. It is now set by the JSON-LD VP builder, JWTs don't have a nonce claim. For that, `jti` is used?

woutslakhorst replied:
In this case the nonce is self generated but in other cases the nonce has to come from the auth server. So it should remain a parameter?

reinkrul replied:
We can always make it a parameter again when we need it.

reinkrul commented on vcr/credential/resolver.go:
```diff
@@ -73,17 +73,27 @@ func PresentationSigner(presentation vc.VerifiablePresentation) (*did.DID, error
-		var proofs []proof.LDProof
-		if err := presentation.UnmarshalProofValue(&proofs); err != nil {
-			return nil, fmt.Errorf("invalid LD-proof for presentation: %w", err)
+		proof, err := ParseLDProof(presentation)
```
you're right, and it was OK due to the default (which is not OK). I'll remove the default and return an error in that case

woutslakhorst reviewed (commented):
Apart from Gerards comments, the handle looks quite good. The rest of the PR is a bit cloudy since there have been some changes in master that should be merged first.

reinkrul commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -40,43 +39,70 @@ import (
-type serviceToService struct {
-}
+// maxPresentationValidity defines the maximum validity of a presentation.
+const maxPresentationValidity = 10 * time.Second
```
Changed to 5 sec. Actual validAt check does is not implemented by this PR, but part of the existing VP verifier. Skew for JWTs was missing there though, so I added that.

reinkrul commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -143,7 +168,137 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+	}
+
+	nonceStore := r.storageEngine.GetSessionDatabase().GetStore(maxPresentationValidity, "s2s", "nonce")
+	err := nonceStore.Get(nonce, new(bool))
+	if !errors.Is(err, storage.ErrNotFound) {
+		if err != nil {
+			// unable to check nonce
+			return err
+		}
+		return oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: "presentation nonce has already been used",
+		}
+	}
```
It's right, but confusing since good cases should return `storage.ErrNotFound`. So I rewrote it a bit to make it more explicit. I also decided to always write the nonce as used, even if the an error occurs while checking it. So a VP can really only be used once.

reinkrul commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -143,7 +168,137 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+			return err
+		}
+		if proof.Domain != nil {
+			audience = []string{*proof.Domain}
```
Separate issue, I'd say: https://github.com/nuts-foundation/nuts-node/issues/2649

reinkrul commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -143,7 +168,137 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+	return nil
+}
+
+func (r *Wrapper) validatePresentationAudience(presentation vc.VerifiablePresentation, issuer did.DID) error {
```
I think we might need to reorganize stuff anyways, but @woutslakhorst is now working on OpenID4VP stuff in parallel, so not the right moment?

reinkrul commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -115,20 +141,19 @@ func (r Wrapper) RequestAccessToken(ctx context.Context, request RequestAccessTo
-		VPToken:                        []VerifiablePresentation{presentation},
-		PresentationDefinition:         nil,
-		PresentationSubmission:         nil,
+		ClientId:               "",
```
It is currently not issued to a specific DID, as there might be multiple credential subject DIDs in the `vp_token`?
I think `client_id` should reflect the authenticated client, but the client doesn't provide a client_id (and certainly not an authenticated client_id).

The authorization code grant's JAR (signed authorization request object), but we don't have that concept here. Ideas;
- specify client_id authentication based on signature with key from DID document
- authenticate client_id using client certificate, which must resolve to did:web being used as client_id
- ...?

woutslakhorst replied:
RFC021 states the following for the introspection:

> The client (DID) that requested the access token.

In the request, the `sub` field indicates the requester. So the DID from that field should be transfered to the client_id.

gerardsn replied:
`sub` == `iss` of the token == resource owner. Sounds like it's missing from the spec. Tokens should be issued to a specific party, and we should be able to check if that party is trusted. 

For now we expect a single VP in the s2s flow and all credentials in a VP must have the same subject == `client_id`?

gerardsn replied:
`sub` == `iss` of the token == resource owner. Sounds like it's missing from the spec. Tokens should be issued to a specific party, and we should be able to check if that party is trusted. 

For now we expect a single VP in the s2s flow and all credentials in a VP must have the same subject == `client_id`?

gerardsn replied:
In our context I think the `client_id` should always be a DID. In OpenID4VP the `client_id_scheme ` can be used to specify a DID as `client_id` in the Authorization Request. Not sure how to use that information, just thought it could be useful in this discussion.

maybe create a new issue and point to it in a TODO for now? IMO it's a blocking issue for production if we can't assign the token to a specific DID.

reinkrul replied:
Just took first VP signer for now, needs to be addressed in RFC first (https://github.com/nuts-foundation/nuts-specification/issues/269)

woutslakhorst commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -143,7 +173,147 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+	if nonceError != nil && errors.Is(nonceError, storage.ErrNotFound) {
+		// this is OK, nonce has not been used before
+		nonceError = nil
+	} else if nonceError != nil {
+		// other error occurred. Keep error to report after storing nonce
+	} else {
+		// no store error: value was retrieved from store, meaning the nonce has been used before
+		nonceError = oauth.OAuth2Error{
+			Code:        oauth.InvalidRequest,
+			Description: "presentation nonce has already been used",
+		}
+	}
```
```
} else if nonceError == nil {
...
}
// else other error occurred. Keep error to report after storing nonce 
```

reinkrul commented on vcr/holder/wallet.go:
```diff
@@ -118,12 +119,16 @@ func (h wallet) buildJWTPresentation(ctx context.Context, subjectDID did.DID, cr
 		jwt.IssuerKey:  subjectDID.String(),
 		jwt.SubjectKey: subjectDID.String(),
 		jwt.JwtIDKey:   id.String(),
+		"nonce":        crypto.GenerateNonce(),
```
challenge is not the same as nonce? I'd say nonce is a technical random value included in the signature to prevent replay attacks, while challenge is a functional value (like the IRMA contract message)? In JSON-LD Proofs they're 2 different fields anyways.

woutslakhorst replied:
openid4vp requires it to be passed

reinkrul commented on vcr/credential/util.go:
```diff
@@ -0,0 +1,110 @@
+	"time"
+)
+
+// ResolveSubjectDID resolves the subject DID from the given credentials.
+// It returns an error if:
+// - the credentials do not have the same subject DID.
+// - the credentials do not have a subject DID.
+func ResolveSubjectDID(credentials ...vc.VerifiableCredential) (*did.DID, error) {
+	var subjectID did.DID
+	for _, credential := range credentials {
+		sid, err := credential.SubjectDID()
+		if err != nil {
+			return nil, err
+		}
+		if !subjectID.Empty() && !subjectID.Equals(*sid) {
+			return nil, errors.New("not all VCs have the same credentialSubject.id")
+		}
+		subjectID = *sid
+	}
+	return &subjectID, nil
+}
+
+// PresenterIsCredentialSubject checks if the presenter of the VP is the same as the subject of the VCs being presented.
+// If the presentation signer or credential subject can't be resolved, it returns an error.
+// If parsing succeeds and the signer DID is the same as the credential subject DID, it returns true.
+func PresenterIsCredentialSubject(vp vc.VerifiablePresentation) (bool, error) {
```
Not sure that will add anything, when you feed the Nuts node a Verifiable Credential that it was never meant to understand (e.g. issuer is not a DID), things will also horribly fail. They're util functions that exactly state what they do, so either call 'm or don't?
If it's important to have a clear disctinction between what is standardized and what not (could be useful), we could consider organization code in such way (e.g. defining things like "profile")

reinkrul commented on auth/client/iam/client.go:
```diff
@@ -137,11 +137,12 @@ func (hb HTTPClient) AccessToken(ctx context.Context, tokenEndpoint string, vp v
 
 	// create a POST request with x-www-form-urlencoded body
-	assertion, _ := json.Marshal(vp)
+	assertion := vp.Raw()
```
I don't think I understand the problem; `assertion` must be an encoded signed VP, meaning a string for JWT (compact form) or JSON object (JSON-LD). When you construct a VP yourself and don't sign it, it is not encoded, so `Raw()` is obviously invalid (the fact that you can call it, might be a shortcoming of go-did).

Marshalling/unmarshalling an empty VP in `Raw()` only masks bugs: you guessing which encoding format (JTW/JSON-LD) and leads a an invalid VP anyways.

gerardsn commented on vcr/credential/util.go:
```diff
@@ -0,0 +1,110 @@
+
+// PresentationIssuanceDate returns the date at which the presentation was issued.
+// For JSON-LD, it looks at the first LinkedData proof's 'created' property.
+// For JWT, it looks at the 'iat' claim, or if that is not present, the 'nbf' claim.
```
I guess allowing a fallback is fine. It should still default to the specced value (`nbf`) and only use the fallback if that is not present though

gerardsn commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -143,7 +168,137 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+	}
+}
+
+func (r *Wrapper) s2sAccessTokenStore() storage.SessionStore {
 	return r.storageEngine.GetSessionDatabase().GetStore(accessTokenValidity, "s2s", "accesstoken")
```
and this one

reinkrul replied:
Maybe. Probably depends on how @woutslakhorst organizes it for the authorization code and OpenID4VP flow?

reinkrul commented on auth/api/iam/s2s_vptoken.go:
```diff
@@ -143,7 +173,141 @@ func (r Wrapper) createAccessToken(issuer did.DID, issueTime time.Time, presenta
+		}
+	case vc.JSONLDPresentationProofFormat:
+		proof, err := credential.ParseLDProof(presentation)
+		if err != nil || proof.Nonce == nil {
```
Well, you can use an empty string only once, so not really an attack vector? But I'll add a check here.
